Reland "[lldb] Introduce ScriptedFrameProvider for real threads (#161870)"

medismailben opened the pull request:
This patch re-lands #161870 with fixes to the previous test failures.

rdar://161834688

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben commented:
This depends on #170191. After that lands, it should only contain one commit.

JDevlieghere reviewed (approved):
I reviewed the original PR so this (still) LGTM.

medismailben committed:
Reland "[lldb] Introduce ScriptedFrameProvider for real threads (#161870)"

This patch extends ScriptedFrame to work with real (non-scripted)
threads,
enabling frame providers to synthesize frames for native processes.

Previously, ScriptedFrame only worked within
ScriptedProcess/ScriptedThread
contexts. This patch decouples ScriptedFrame from ScriptedThread,
allowing
users to augment or replace stack frames in real debugging sessions for
use
cases like custom calling conventions, reconstructing corrupted frames
from
core files, or adding diagnostic frames.

Key changes:

- ScriptedFrame::Create() now accepts ThreadSP instead of requiring
ScriptedThread, extracting architecture from the target triple rather
than ScriptedProcess.arch

- Added SBTarget::RegisterScriptedFrameProvider() and
ClearScriptedFrameProvider() APIs, with Target storing a
SyntheticFrameProviderDescriptor template for new threads

- Added "target frame-provider register/clear" commands for CLI access

- Thread class gains LoadScriptedFrameProvider(),
ClearScriptedFrameProvider(),
and GetFrameProvider() methods for per-thread frame provider management

- New SyntheticStackFrameList overrides FetchFramesUpTo() to lazily
provide
frames from either the frame provider or the real stack

This enables practical use of the SyntheticFrameProvider infrastructure
in
real debugging workflows.

rdar://161834688

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

google-yfyang commented:
Looks like we are seeing some test failures similar to ones reported in the previous attempt. 

Can you take a look at this?
 
```
FAIL: LLDB (
INTERNAL_DIRECTORY/lldb/testing/clang-x86_64) :: test_scripted_frame_objects (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_scripted_frame_objects)
<bound method SBProcess.Kill of SBProcess: pid = 723, state = exited, threads = 3, executable = a.out>: success

Restore dir to: INTERNAL_DIRECTORY
======================================================================
FAIL: test_append_frames (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_append_frames)
   Test that we can add frames after real stack.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "INTERNAL_DIRECTORY/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 122, in test_append_frames
    self.assertEqual(new_frame_count, original_frame_count + 1)
AssertionError: 6 != 7
Config=x86_64-/INTERNAL_DIRECTORY/lldb/testing/clang
======================================================================
FAIL: test_applies_to_thread (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_applies_to_thread)
   Test that applies_to_thread filters which threads get the provider.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/INTERNAL_DIRECTORY/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 218, in test_applies_to_thread
    self.assertEqual(
AssertionError: 6 != 1 : Thread with ID 1 should have 1 synthetic frame
Config=x86_64-/INTERNAL_DIRECTORY/lldb/testing/clang
======================================================================
FAIL: test_circular_dependency_fix (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_circular_dependency_fix)
   Test that accessing input_frames in __init__ doesn't cause circular dependency.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/INTERNAL_DIRECTORY/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 388, in test_circular_dependency_fix
    self.assertEqual(
AssertionError: 6 != 7 : Should have original frames + 1 synthetic frame
Config=x86_64-/INTERNAL_DIRECTORY/lldb/testing/clang
======================================================================
FAIL: test_prepend_frames (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_prepend_frames)
   Test that we can add frames before real stack.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/INTERNAL_DIRECTORY/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 84, in test_prepend_frames
    self.assertEqual(new_frame_count, original_frame_count + 2)
AssertionError: 6 != 8
Config=x86_64-/INTERNAL_DIRECTORY/lldb/testing/clang
======================================================================
FAIL: test_remove_frame_provider_by_id (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_remove_frame_provider_by_id)
   Test that RemoveScriptedFrameProvider removes a specific provider by ID.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/INTERNAL_DIRECTORY/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 272, in test_remove_frame_provider_by_id
    self.assertEqual(thread.GetNumFrames(), 3, "Should have 3 synthetic frames")
AssertionError: 6 != 3 : Should have 3 synthetic frames
Config=x86_64-/INTERNAL_DIRECTORY/lldb/testing/clang
======================================================================
FAIL: test_replace_all_frames (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_replace_all_frames)
   Test that we can replace the entire stack.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/INTERNAL_DIRECTORY/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 41, in test_replace_all_frames
    self.assertEqual(thread.GetNumFrames(), 3, "Should have 3 synthetic frames")
AssertionError: 6 != 3 : Should have 3 synthetic frames
Config=x86_64-/INTERNAL_DIRECTORY/lldb/testing/clang
======================================================================
FAIL: test_scripted_frame_objects (TestScriptedFrameProvider.ScriptedFrameProviderTestCase.test_scripted_frame_objects)
   Test that provider can return ScriptedFrame objects.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/INTERNAL_DIRECTORY/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 152, in test_scripted_frame_objects
    self.assertEqual(
AssertionError: 6 != 5 : Should have 5 custom scripted frames
Config=x86_64-/INTERNAL_DIRECTORY/lldb/testing/clang
----------------------------------------------------------------------
Ran 7 tests in 13.556s
```
